src/toil/lib/ec2.py

```diff
@@ -530,3 +536,67 @@ def create_auto_scaling_group(autoscaling_client: BaseClient,
 
     # Don't prune the ASG because MinSize and MaxSize are required and may be 0.
     autoscaling_client.create_auto_scaling_group(**asg)
+
+
+def get_flatcar_ami(ec2_client: BaseClient) -> str:
+    """
+    Retrieve the flatcar AMI image to use as the base for all Toil autoscaling instances.
+
+    AMI must be available to the user on AWS (attempting to launch will return a 403 otherwise).
+
+    Priority is:
+      1. User specified AMI via TOIL_AWS_AMI
+      2. Official AMI from stable.release.flatcar-linux.net
+      3. Search the AWS Marketplace
+
+    If all of these sources fail, we raise an error to complain.
+    """
+    # Take a user override
+    ami = os.environ.get('TOIL_AWS_AMI')
+    if not ami:
+        ami = official_flatcar_ami_release(ec2_client=ec2_client)
+    if not ami:
+        ami = aws_marketplace_flatcar_ami_search(ec2_client=ec2_client)
+    if not ami:
+        raise RuntimeError('Unable to fetch the latest flatcar image.')
+    return ami
+
+
+@retry()  # TODO: What errors do we get for timeout, JSON parse failure, etc?
+def official_flatcar_ami_release(ec2_client: BaseClient) -> Optional[str]:
+    """Check stable.release.flatcar-linux.net for the latest flatcar AMI.  Verify it's on AWS."""
+    # Flatcar images only live for 9 months.
+    # Rather than hardcode a list of AMIs by region that will die, we use
+    # their JSON feed of the current ones.
+    JSON_FEED_URL = 'https://stable.release.flatcar-linux.net/amd64-usr/current/flatcar_production_ami_all.json'
+    region = ec2_client._client_config.region_name
+    feed = json.loads(urllib.request.urlopen(JSON_FEED_URL).read())
+
+    try:
+        for ami_record in feed['amis']:
+            # Scan the klist of regions
+            if ami_record['name'] == region:
+                # When we find ours, return the AMI ID
+                ami = ami_record['hvm']
+                # verify it exists on AWS
+                response = ec2_client.describe_images(Filters=[{'Name': 'image-id', 'Values': [ami]}])
+                if len(response['Images']) == 1 and response['Images'][0]['State'] == 'available':
+                    return ami
+        # We didn't find it
+        logger.warning(f'Flatcar image feed at {JSON_FEED_URL} does not have an image for region {region}')
+    except KeyError:
+        # We didn't see a field we need
+        logger.warning(f'Flatcar image feed at {JSON_FEED_URL} does not have expected format')
+
+
+@retry()  # TODO: What errors do we get for timeout, JSON parse failure, etc?
+def aws_marketplace_flatcar_ami_search(ec2_client: BaseClient) -> Optional[str]:
+    """Query AWS for all AMI names matching 'Flatcar-stable-*' and return the most recent one."""
+    response: dict = ec2_client.describe_images(Owners=['aws-marketplace'],
+                                                Filters=[{'Name': 'name', 'Values': ['Flatcar-stable-*']}])
```

Comment:
Are you sure this won't start up the "Premium" Flatcar AMIs? And is it clever enough to actually subscribe to the marketplace product? I think we manually subscribed the Toil account.

Reply:
I'm not sure if it'll subscribe, I thought that was only `coreos`, though I could have just forgotten.

The pro images have the pattern `Flatcar-pro-stable-*`.  The others are `Flatcar-alpha-*`, `Flatcar-beta-*`, and `Flatcar-edge-*`.

Reply:
I couldn't find an easy way to subscribe programmatically.  That doesn't mean there isn't one, but either way, it might need some digging to find out and is probably appropriate for a separate PR if it's possible.